File: y15/d05/cpp/main.cpp
#include "helper.h"
#include <iostream>
#include <ostream>
#include <regex>
#include <vector>
using namespace std;

int main(int argc, char *argv[]) {
  if (argc < 2) {
    cout << "Provide the input file name" << endl;
    return 1;
  }

  vector<string> lines = read_lines(argv[1]);
  // Part 1
  int valid_count1 = 0;
  for (const auto &line : lines) {
    // regex pattern("\\d+");
    vector<string> vowels = re_match(line, regex("(a|e|i|o|u)"));
    if (vowels.size() < 3)
      continue;
    vector<string> twice = re_match(line, regex("(\\w)\\1"));
    if (twice.size() == 0)
      continue;
    vector<string> bad_strs = re_match(line, regex("(ab|cd|pq|xy)"));
    if (bad_strs.size() > 0)
      continue;
    valid_count1++;
  }
  cout << "Part 1: " << valid_count1 << endl;

  // Part 2
  int valid_count2 = 0;
  for (const auto &line : lines) {
    // regex pattern("\\d+");
    vector<string> pairs = re_match(line, regex("(\\w{2}).*\\1"));
    if (pairs.size() == 0)
      continue;
    vector<string> onebetween = re_match(line, regex("(\\w)\\w\\1"));
    if (onebetween.size() == 0)
      continue;
    valid_count2++;
  }
  cout << "Part 2: " << valid_count2 << endl;


  return 0;
}

File: y15/d05/rust/src/main.rs
use regex::Regex;
use std::fs::read_to_string;
use std::{char, env};

fn main() {
    let lines: Vec<String> = read_to_string(env::args().nth(1).expect("Error: Missing file name"))
        .unwrap()
        .lines()
        .map(String::from)
        .collect();
    solve(lines);
}

fn solve(lines: Vec<String>) {
    // PART 1
    let re_vowels = Regex::new(r"(a|e|i|o|u)").unwrap();
    let re_bad = Regex::new(r"(ab|cd|pq|xy)").unwrap();
    let mut count1 = 0;
    for line in &lines {
        let vowels = re_vowels.find_iter(line);
        if vowels.count() < 3 {
            continue;
        }

        // check for double
        let mut chars = line.chars();
        let mut check = chars.next();
        let mut doubles = false;
        for curc in chars {
            if check == Some(curc) {
                doubles = true;
                break;
            }
            check = Some(curc)
        }
        if !doubles {
            continue;
        }

        let bad = re_bad.find_iter(line);
        if bad.count() > 0 {
            continue;
        }

        count1 += 1;
    }

    println!("Part 1: {}", count1);

    // PART 2
    let mut count2 = 0;
    for line in &lines {
        let mut pairs = false;
        let chars: Vec<char> = line.chars().collect();

        for i in 0..chars.len() - 3 {
            for j in i + 2..chars.len() - 1 {
                if chars[i] == chars[j] && chars[i + 1] == chars[j + 1] {
                    pairs = true;
                    break;
                }
            }
        }

        if !pairs {
            continue;
        }

        let mut one_between = false;
        for i in 0..chars.len() - 2 {
            if chars[i] == chars[i + 2] {
                one_between = true;
                break;
            }
        }

        if !one_between {
            continue;
        }

        count2 += 1;
    }

    println!("Part 2 {}", count2);
}
